src/number_theory/fermat4.lean

```diff
@@ -0,0 +1,322 @@
+/-
+Copyright (c) 2020 Paul van Wamelen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Paul van Wamelen.
+-/
+import number_theory.pythagorean_triples
+
+/-!
+# Fermat's Last Theorem for the case n = 4
+There are no non-zero integers `a`, `b` and `c` such that `a ^ 4 + b ^ 4 = c ^ 4`.
+-/
+
+noncomputable theory
+open_locale classical
+
+/-- Shorthand for three non-zero integers `a`, `b`, and `c` satisfying `a ^ 4 + b ^ 4 = c ^ 2`.
+We will show that no integers satisfy this equation. Clearly Fermat's Last theorem for n = 4
+follows. -/
+def fermat_42 (a b c : ℤ) : Prop := a ≠ 0 ∧ b ≠ 0 ∧ c ≠ 0 ∧ a ^ 4 + b ^ 4 = c ^ 2
+
+lemma fermat_42_comm {a b c : ℤ} :
+ (fermat_42 a b c) ↔ (fermat_42 b a c) :=
+ by { delta fermat_42, rw add_comm, tauto }
+
+lemma fermat_42_mul {a b c k : ℤ} (hk0 : k ≠ 0) :
+  fermat_42 a b c ↔ fermat_42 (k * a) (k * b) (k ^ 2 * c) :=
+begin
+  delta fermat_42,
+  split,
+  { intro f42,
+    split, { exact mul_ne_zero hk0 f42.1 },
+    split, { exact mul_ne_zero hk0 f42.2.1 },
+    split, { apply mul_ne_zero, { rw pow_two, exact mul_ne_zero hk0 hk0 }, exact f42.2.2.1 },
+    { calc (k * a) ^ 4 + (k * b) ^ 4
+          = k ^ 4 * (a ^ 4 + b ^ 4) : by ring
+      ... = k ^ 4 * c ^ 2 : by rw f42.2.2.2
+      ... = (k ^ 2 * c) ^ 2 : by ring }},
+  { intro f42,
+    split, { exact right_ne_zero_of_mul f42.1 },
+    split, { exact right_ne_zero_of_mul f42.2.1 },
+    split, { exact right_ne_zero_of_mul f42.2.2.1 },
+    apply (mul_right_inj' (pow_ne_zero 4 hk0)).mp,
+    { calc k ^ 4 * (a ^ 4 + b ^ 4)
+          = (k * a) ^ 4 + (k * b) ^ 4 : by ring
+      ... = (k ^ 2 * c) ^ 2 : by rw f42.2.2.2
+      ... = k ^ 4 * c ^ 2 : by ring }}
+end
+
+/-- We say a solution to `a ^ 4 + b ^ 4 = c ^ 2` is minimal if there are no other solution with
+a smaller c (in absolute value). -/
+def minimal_fermat_42 (a b c : ℤ) : Prop :=
+  (fermat_42 a b c) ∧ ∀ (a1 b1 c1 : ℤ), (fermat_42 a1 b1 c1) → int.nat_abs c ≤ int.nat_abs c1
```

Comment:
```suggestion
def fermat_42.minimal {a b c : ℤ} (fermat_42 a b c) : Prop :=
∀ {{a1 b1 c1 : ℤ}}, (fermat_42 a1 b1 c1) → int.nat_abs c ≤ int.nat_abs c1
```
This way, if you have `h : fermat_42 a b c`, then you can write `h.minimal` for the assertion that the triple is minimal.
This might allow you to slightly compress some of the statements/proofs. But it's not very important, because it will not be used outside this file (I guess).

Reply:
I tried this, but it seems to make later statements less clear.
```
lemma exists_minimal {a b c : ℤ} (h : fermat_42 a b c) :
  ∃ (a0 b0 c0), (minimal a0 b0 c0)
```
becomes
```
lemma exists_minimal {a b c : ℤ} (h : fermat_42 a b c) :
  ∃ (a0 b0 c0) (hf : fermat_42 a0 b0 c0), hf.minimal
```
And
```
lemma minimal_comm {a b c : ℤ} :
 (minimal a b c) → (minimal b a c) :=
```
becomes
```
lemma minimal_comm {a b c : ℤ} (h1 : fermat_42 a b c) (hm1 : h1.minimal) :
 (fermat_42.comm.mp h1).minimal
```
What do you think? The original seems clearer to me.

Reply:
Ok, let's leave it like this.